### PR TITLE
Fix for NumPy 1.23

### PR DESCRIPTION
**Pull request by takagi**
Rel #6764.

### CODE REVIEW

**Comment by toslunar** [comment]
For CuPy v11, I'd follow NumPy 1.23's behavior by removing these lines
https://github.com/cupy/cupy/blob/16b511fd3f24b8e174d7c61198a1b5f81c46425c/cupy/_core/_routines_indexing.pyx#L259-L270

**Comment by takagi** [comment]
Sure! (I completely had misunderstanding about that...)

File: tests/cupy_tests/core_tests/test_ndarray_adv_indexing.py [hunk 1]
```diff
@@ -146,35 +150,33 @@ def test_adv_getitem(self, xp, dtype):
...
-            return a[self.indexes]
+        a = testing.shaped_arange(self.shape, xp, dtype)
+        return a[self.indexes]
 
 
 @testing.parameterize(
-    {'shape': (0,), 'indexes': True},
-    {'shape': (0,), 'indexes': (True,)},
-    {'shape': (0,), 'indexes': numpy.ones((), dtype=numpy.bool_)},
-    {'shape': (0,), 'indexes': numpy.zeros((), dtype=numpy.bool_)},
```
Comment (takagi) [inline tests/cupy_tests/core_tests/test_ndarray_adv_indexing.py:168]: These tests were split from `TestArrayAdvancedIndexingGetitemParametrized` in #1503 for NumPy 1.13 support, so I've made them back there in this PR.

File: tests/cupy_tests/core_tests/test_ndarray_adv_indexing.py [hunk 2]
```diff
@@ -538,29 +543,8 @@ def test_adv_setitem(self, xp, dtype):
...
-        with testing.assert_warns(FutureWarning):
-            a[self.indexes] = self.value
-        return a
-
-
-@testing.parameterize(
-    {'shape': (0,), 'indexes': True, 'value': 1},
-    {'shape': (0,), 'indexes': (True,), 'value': 1},
-    {'shape': (0,), 'indexes': numpy.ones((), dtype=numpy.bool_), 'value': 1},
-    {'shape': (0,), 'indexes': numpy.zeros((), dtype=numpy.bool_), 'value': 1},
```
Comment (takagi) [inline tests/cupy_tests/core_tests/test_ndarray_adv_indexing.py:561]: These tests were split from `TestArrayAdvancedIndexingSetitemScalarValue` in https://github.com/cupy/cupy/pull/1503 for NumPy 1.13 support, so I've made them back there in this PR as numpy<1.13 is dropped.

File: cupyx/jit/_cuda_typerules.py [hunk 3]
```diff
@@ -82,7 +73,7 @@
     ast.Sub: arithmetic.subtract,
     ast.Mult: arithmetic.multiply,
     ast.Pow: arithmetic.power,
-    ast.Div: _numpy_scalar_true_divide,
```
Comment (takagi) [inline cupyx/jit/_cuda_typerules.py:85]: Seems that we don't need `_numpy_scalar_true_divide` anymore.

File: cupy/_core/core.pyx [hunk 4]
```diff
@@ -653,14 +653,6 @@ cdef class _ndarray_base:
...
-        elif self._f_contiguous:
-            warnings.warn(
-                'Changing the shape of an F-contiguous array by '
-                'descriptor assignment is deprecated. To maintain the '
-                'Fortran contiguity of a multidimensional Fortran '
-                'array, use \'a.T.view(...).T\' instead',
-                DeprecationWarning)
-            axis = 0
         else:
             # Don't mention the deprecated F-contiguous support
```
Comment (toslunar) [inline cupy/_core/core.pyx:657]: nit: This line can be removed, too.

File: cupy/_core/internal.pyx [hunk 5]
```diff
@@ -130,14 +130,28 @@ cdef void get_reduced_dims(
...
+        Py_ssize_t itemsize, bint is_c_contiguous, bint zeros_for_zerosize):
     cdef Py_ssize_t st, sh
     cdef Py_ssize_t is_nonzero_size = 1
     cdef int i, ndim = shape.size()
     cdef Py_ssize_t idx
     strides.resize(ndim, 0)
     st = 1
 
+    # For some cases, the strides are all set to zero if the shape is
+    # zero-sized since NumPy 1.23.
```
Comment (toslunar) [inline cupy/_core/internal.pyx:142]: Do you know whether NumPy intentionally kept the behavior for no copy reshape?  Maybe it's good to report to NumPy.

File: cupyx/jit/_cuda_typerules.py (hunk 3 above)
Comment (toslunar) [reply cupyx/jit/_cuda_typerules.py:85]: It seems so to me, too.

@asi1024 Do you have any comment?

**Review by toslunar — APPROVED** [review]
LGTM

I left some minor comments that seem optional to be fixed in this PR.

File: cupy/_core/core.pyx (hunk 4 above)
Comment (takagi) [reply cupy/_core/core.pyx:657]: Sure, thanks!

File: cupy/_core/internal.pyx (hunk 5 above)
Comment (takagi) [reply cupy/_core/internal.pyx:142]: No, I couldn't search the source enough last night. As you say, it's a kind of weird behavior. I'm going to see it again.

Comment (takagi) [reply cupy/_core/internal.pyx:142]: It seems the change in creation (zero cleared strides) was introduced in https://github.com/numpy/numpy/pull/21477/.

**Review by toslunar — APPROVED** [review]

File: cupyx/jit/_cuda_typerules.py (hunk 3 above)
Comment (asi1024) [reply cupyx/jit/_cuda_typerules.py:85]: I agree to use arithmetic.true_divide!

**Comment by toslunar** [comment]
v10 fix will be `@testing.with_requires('numpy<1.23')`.

**Pull request merged**